# Fundamentos de Ciencias de Datos

*Esta tarea tiene dos partes. Una parte en parejas y otra individual. La parte en parejas se encuentra en el archivo **tarea_01.ipynb**. Luego, cada uno de los integrantes de la pareja debe tomar la sección A o la sección B, y cada integrante debe realizar solamente la sección de la tarea que escogió. Este cuaderno de python contiene la realización de la **sección A** de la tarea, parte individual*

## Tarea 1 (Parte Individual - Sección A)

### Objetivos

1. Familiarización con Python en el contexto de Ciencias de Datos
2. Manejo de conjuntos de datos utilizando funciones propias de Pandas

### Metodología

Deberás completar este cuaderno de python, siguiendo las instrucciones contenidas en él. Puedes utilizar contenido que encuentre en internet siempre y cuando lo referencies a través de un comentario. Por ejemplo:

>```Python
># código para abrir archivo obtenido desde  https://uniwebsidad.com/libros/algoritmos-python/capitulo-11
>archivo = open("archivo.txt")
>```

No olvides poner nombres representativos a tus variables. Caso contrario se verá reflejado en una penalización de tu nota.

### Conjuntos de Datos

Se le entregan 3 conjuntos de datos: codigos.csv, datos_comunas.csv, presupuesto_2019.csv. 

#### codigos.csv

 | columna | descripción |
 | --- | :---: |
 | Código Región | **Entero**: Identificador único para cada región. |
 | Nombre Región | **String**: Nombre de la región. |
 | Abreviatura Región | **String**: Abreviatura (en letras) o nombre corto de la región.  |
 | Código Provincia | **Entero**: Identificador único para cada provincia. |
 | Nombre Provincia | **String**: Nombre de la provincia. |
 | Código Comuna 2018 | **Entero**: Identificador único para cada comuna. |
 | Nombre Comuna | **String**: Nombre de la comuna. |

 #### datos_comunas.csv

 | columna | descripción |
 | --- | :---: |
 | id | **Entero**: Identificador único para cada comuna. Mismo identificador que en **codigos.csv**. |
 | metros_plaza | **Entero**: Cantidad de metros de plaza que tiene la comuna. |
 | metros_parque | **Entero**: Cantidad de metros de parque que tiene la comuna.  |
 | personas_trabaja | **Entero**: Número de personas que trabaja en la municipalidad de la comuna. |

 #### presupuesto_2019.csv

 | columna | descripción |
 | --- | :---: |
 | CODIGO | **Entero**: Identificador único para cada comuna. Mismo identificador que en **codigos.csv**. |
 | MUNICIPIO | **String**: Nombre de la comuna. |
 | PRESUPUESTO | **Entero**: Presupuesto destinado a la comuna.  |

 **Observación**: en la columna `PRESUPUESTO`, hay datos que están como "no recepcionados. Eso puede producir que los análisis fallen, por lo que se recomienda limpiar los datos antes.

 #### poblacion.csv

 | columna | descripción |
 | --- | :---: |
 | CODIGO | **Entero**: Identificador único para cada comuna. Mismo identificador que en **codigos.csv**. |
 | MUNICIPIO | **String**: Nombre de la comuna. |
 | POBLACIÓN | **Entero**: Número de personas que vive en la comuna.  |



## Nombre alumno

Por favor, llenar acá con su nombre.

**José Facundo Barrientos Juncos**


## Recomendaciones

Primero, deberías importar en la siguiente celda **TODOS** los módulos de Python necesarios para la correcta ejecución de su tarea.

In [ ]:
import pandas as pd
import numpy as np

Luego, deberías guardar en una (o varias) variable(s) TODOS los sets de datos que utilizarás en la tarea.

In [ ]:
df_codigos = pd.read_csv("codigos.csv", sep=";")
df_datos_comuna = pd.read_csv("datos_comunas.csv")
df_poblacion = pd.read_csv("poblacion.csv", sep=";")
df_presupuesto = pd.read_csv("presupuesto_2019.csv", sep=";")

Utilice las variables previamente guardadas para mostrar parte de la información de todos los conjuntos de datos. Vale decir, si usted creó dos variables, cada una con un conjunto de datos diferente, muestre parte de los dos conjuntos de datos. Esto te ayudará a entender mejor lo que contiene cada fuente de datos.

In [ ]:
df_codigos

Código Región Nombre Región  ... Código Comuna 2018  Nombre Comuna
0                1      Tarapacá  ...               1101        Iquique
1                1      Tarapacá  ...               1107  Alto Hospicio
2                1      Tarapacá  ...               1401   Pozo Almonte
3                1      Tarapacá  ...               1402         Camiña
4                1      Tarapacá  ...               1403       Colchane
..             ...           ...  ...                ...            ...
341             16         Ñuble  ...              16301     San Carlos
342             16         Ñuble  ...              16302       Coihueco
343             16         Ñuble  ...              16303         Ñiquén
344             16         Ñuble  ...              16304     San Fabián
345             16         Ñuble  ...              16305    San Nicolás

[346 rows x 7 columns]

In [ ]:
df_datos_comuna

id  metros_plaza  metros_parque  personas_trabaja
0     1101       58460.0        94824.0             271.0
1     1107      130316.0            0.0              55.0
2     1401           NaN            NaN               NaN
3     1402        4050.0            0.0               8.0
4     1403        2532.0            NaN               NaN
..     ...           ...            ...               ...
340  16301      168800.0        82226.0              19.0
341  16302       21282.0            0.0              62.0
342  16303       21925.0            0.0              22.0
343  16304        9500.0            0.0              26.0
344  16305       36300.0         4180.0              33.0

[345 rows x 4 columns]

In [ ]:
df_poblacion

CODIGO      MUNICIPIO  POBLACIÓN
0      1101        IQUIQUE     216514
1      1107  ALTO HOSPICIO     124150
2      1401   POZO ALMONTE      16683
3      1402         CAMIÑA       1345
4      1403       COLCHANE       1556
..      ...            ...        ...
340   16301     SAN CARLOS      55935
341   16302       COIHUECO      28147
342   16303         ÑIQUÉN      11556
343   16304     SAN FABIÁN       4607
344   16305    SAN NICOLÁS      12068

[345 rows x 3 columns]

In [ ]:
df_presupuesto

CODIGO      MUNICIPIO PRESUPUESTO
0      1101        IQUIQUE    59072234
1      1107  ALTO HOSPICIO    13984411
2      1401   POZO ALMONTE     7613962
3      1402         CAMIÑA     1584008
4      1403       COLCHANE     2500000
..      ...            ...         ...
340   16301     SAN CARLOS    10459691
341   16302       COIHUECO     5001190
342   16303         ÑIQUÉN     3386698
343   16304     SAN FABIÁN     1798500
344   16305    SAN NICOLÁS     3979700

[345 rows x 3 columns]

Además, si necesitas hacer algún pre-procesamiento **COMÚN** a todas sus futuras respuestas, deberías hacerlo en la siguiente celda

In [ ]:
#Escriba su código aquí

## Actividad

### Parte Individual A
1. Genera un dataframe ordenado de manera descendente (más alta primero, más baja al final) según la relación $\frac{\text{metros de plaza}}{\text{población}}$. Las columnas que debes mostrar son el nombre de la comuna y la proporción señalada.
2. Calcula el presupuesto promedio por región. ¿Cuál es la región que tiene el mayor presupuesto promedio?
3. Con base en los datos, responda la siguiente pregunta: revise las 20 comunas con mayor cantidad de población. ¿Cuáles de ellas también son parte de las comunas con más presupuesto? ¿Qué puede comentar de esta situación?

## Detalles académicos

La nota de cada integrante de la pareja será calculado como $0.5 \cdot \text{Parte Grupal} + 0.5 \cdot \text{Parte Individual} $, donde la Parte Grupal corresponde a las 7 primeras consultas, y la parte individual a la sección respectiva elegida por cada integrante. **La tarea debe ser subida al buzón correspondiente de Webcursos hasta el XX de Agosto hasta las XX:XX horas**.

**Importante**: Recuerde que debe generar el código que responda las preguntas. Por ejemplo para la pregunta 4, debe entregar el código que muestre por pantalla la región con la mayor cantidad de metros de plaza. Si entrega sólo el dataframe ordenado su respuesta **no estará completamente correcta**.

## Pregunta 1

In [148]:
df_datosxpoblacion = df_datos_comuna.merge(df_poblacion, left_on="id", right_on="CODIGO")
df_datosxpoblacion

id  metros_plaza  metros_parque  ...  CODIGO      MUNICIPIO POBLACIÓN
0     1101       58460.0        94824.0  ...    1101        IQUIQUE    216514
1     1107      130316.0            0.0  ...    1107  ALTO HOSPICIO    124150
2     1401           NaN            NaN  ...    1401   POZO ALMONTE     16683
3     1402        4050.0            0.0  ...    1402         CAMIÑA      1345
4     1403        2532.0            NaN  ...    1403       COLCHANE      1556
..     ...           ...            ...  ...     ...            ...       ...
340  16301      168800.0        82226.0  ...   16301     SAN CARLOS     55935
341  16302       21282.0            0.0  ...   16302       COIHUECO     28147
342  16303       21925.0            0.0  ...   16303         ÑIQUÉN     11556
343  16304        9500.0            0.0  ...   16304     SAN FABIÁN      4607
344  16305       36300.0         4180.0  ...   16305    SAN NICOLÁS     12068

[345 rows x 7 columns]

In [ ]:
df_datosxpoblacion = df_datosxpoblacion.dropna()
df_datosxpoblacion

id  metros_plaza  metros_parque  ...  CODIGO      MUNICIPIO POBLACIÓN
0     1101       58460.0        94824.0  ...    1101        IQUIQUE    216514
1     1107      130316.0            0.0  ...    1107  ALTO HOSPICIO    124150
3     1402        4050.0            0.0  ...    1402         CAMIÑA      1345
5     1404        2300.0            0.0  ...    1404          HUARA      2914
6     1405       12925.0        20749.0  ...    1405           PICA      5744
..     ...           ...            ...  ...     ...            ...       ...
340  16301      168800.0        82226.0  ...   16301     SAN CARLOS     55935
341  16302       21282.0            0.0  ...   16302       COIHUECO     28147
342  16303       21925.0            0.0  ...   16303         ÑIQUÉN     11556
343  16304        9500.0            0.0  ...   16304     SAN FABIÁN      4607
344  16305       36300.0         4180.0  ...   16305    SAN NICOLÁS     12068

[307 rows x 7 columns]

In [ ]:
df_datosxpoblacion = df_datosxpoblacion[["MUNICIPIO","POBLACIÓN","metros_plaza"]]
df_datosxpoblacion

MUNICIPIO  POBLACIÓN  metros_plaza
0          IQUIQUE     216514       58460.0
1    ALTO HOSPICIO     124150      130316.0
3           CAMIÑA       1345        4050.0
5            HUARA       2914        2300.0
6             PICA       5744       12925.0
..             ...        ...           ...
340     SAN CARLOS      55935      168800.0
341       COIHUECO      28147       21282.0
342         ÑIQUÉN      11556       21925.0
343     SAN FABIÁN       4607        9500.0
344    SAN NICOLÁS      12068       36300.0

[307 rows x 3 columns]

In [ ]:
df_datosxpoblacion.columns= ["Comuna","Poblacion","Metros_de_plaza"]
df_datosxpoblacion

Comuna  Poblacion  Metros_de_plaza
0          IQUIQUE     216514          58460.0
1    ALTO HOSPICIO     124150         130316.0
3           CAMIÑA       1345           4050.0
5            HUARA       2914           2300.0
6             PICA       5744          12925.0
..             ...        ...              ...
340     SAN CARLOS      55935         168800.0
341       COIHUECO      28147          21282.0
342         ÑIQUÉN      11556          21925.0
343     SAN FABIÁN       4607           9500.0
344    SAN NICOLÁS      12068          36300.0

[307 rows x 3 columns]

In [ ]:
df_datosxpoblacion = df_datosxpoblacion.assign(MdP_P=lambda x: x.Metros_de_plaza/x.Poblacion)
df_datosxpoblacion

Comuna  Poblacion  Metros_de_plaza     MdP_P
0          IQUIQUE     216514          58460.0  0.270006
1    ALTO HOSPICIO     124150         130316.0  1.049666
3           CAMIÑA       1345           4050.0  3.011152
5            HUARA       2914           2300.0  0.789293
6             PICA       5744          12925.0  2.250174
..             ...        ...              ...       ...
340     SAN CARLOS      55935         168800.0  3.017789
341       COIHUECO      28147          21282.0  0.756102
342         ÑIQUÉN      11556          21925.0  1.897283
343     SAN FABIÁN       4607           9500.0  2.062079
344    SAN NICOLÁS      12068          36300.0  3.007955

[307 rows x 4 columns]

In [ ]:
df_datosxpoblacion = df_datosxpoblacion[["Comuna","MdP_P"]]
df_datosxpoblacion

Comuna     MdP_P
0          IQUIQUE  0.270006
1    ALTO HOSPICIO  1.049666
3           CAMIÑA  3.011152
5            HUARA  0.789293
6             PICA  2.250174
..             ...       ...
340     SAN CARLOS  3.017789
341       COIHUECO  0.756102
342         ÑIQUÉN  1.897283
343     SAN FABIÁN  2.062079
344    SAN NICOLÁS  3.007955

[307 rows x 2 columns]

In [ ]:
df_datosxpoblacion = df_datosxpoblacion.sort_values("MdP_P", ascending= False)
df_datosxpoblacion

Comuna        MdP_P
50       RINCONADA  3451.165577
252      PRIMAVERA    24.831897
245     RÍO IBÁÑEZ    17.716995
242      O´HIGGINS    16.903817
67   SANTO DOMINGO    16.885927
..             ...          ...
115          MAULE     0.040019
36      COMBARBALÁ     0.032529
27       ANDACOLLO     0.006479
61         NOGALES     0.005266
181        LAUTARO     0.002477

[307 rows x 2 columns]

Obtenida esta tabla se puede decir que las 3 comunas que poseen una mayor proporción entre los metros de plaza y la población son: Rinconada con 3451.166, Primavera con 24.832 y Río Ibañez con 17.717. Y las comunas con una menor proporción son: Lautaro con 0.002, Nogales con 0.005 y Andacollo con 0.006.

## Pregunta 2

In [139]:
df_presupuesto = df_presupuesto.replace('No Recepcionado', 0)
df_presupuesto["PRESUPUESTO"] = df_presupuesto["PRESUPUESTO"].astype(int)
df_presupuesto.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 345 entries, 0 to 344
Data columns (total 3 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   CODIGO       345 non-null    int64 
 1   MUNICIPIO    345 non-null    object
 2   PRESUPUESTO  345 non-null    int64 
dtypes: int64(2), object(1)
memory usage: 8.2+ KB


In [140]:
df_presxcod = df_presupuesto.merge(df_codigos, left_on="CODIGO", right_on="Código Comuna 2018")
df_presxcod

CODIGO      MUNICIPIO  ...  Código Comuna 2018  Nombre Comuna
0      1101        IQUIQUE  ...                1101        Iquique
1      1107  ALTO HOSPICIO  ...                1107  Alto Hospicio
2      1401   POZO ALMONTE  ...                1401   Pozo Almonte
3      1402         CAMIÑA  ...                1402         Camiña
4      1403       COLCHANE  ...                1403       Colchane
..      ...            ...  ...                 ...            ...
340   16301     SAN CARLOS  ...               16301     San Carlos
341   16302       COIHUECO  ...               16302       Coihueco
342   16303         ÑIQUÉN  ...               16303         Ñiquén
343   16304     SAN FABIÁN  ...               16304     San Fabián
344   16305    SAN NICOLÁS  ...               16305    San Nicolás

[345 rows x 10 columns]

In [141]:
df_presxcod = df_presxcod[["Nombre Región","PRESUPUESTO"]]
df_presxcod

Nombre Región  PRESUPUESTO
0        Tarapacá     59072234
1        Tarapacá     13984411
2        Tarapacá      7613962
3        Tarapacá      1584008
4        Tarapacá      2500000
..            ...          ...
340         Ñuble     10459691
341         Ñuble      5001190
342         Ñuble      3386698
343         Ñuble      1798500
344         Ñuble      3979700

[345 rows x 2 columns]

In [142]:
df_presxcod.columns = ["Region","Presupuesto"]
df_presxcod

Region  Presupuesto
0    Tarapacá     59072234
1    Tarapacá     13984411
2    Tarapacá      7613962
3    Tarapacá      1584008
4    Tarapacá      2500000
..        ...          ...
340     Ñuble     10459691
341     Ñuble      5001190
342     Ñuble      3386698
343     Ñuble      1798500
344     Ñuble      3979700

[345 rows x 2 columns]

In [144]:
df_presxcod["Presupuesto"].groupby([df_presxcod['Region']]).mean().sort_values(ascending=False)

Region
Metropolitana de Santiago                    4.341656e+07
Antofagasta                                  2.123443e+07
Tarapacá                                     1.351320e+07
Coquimbo                                     1.351013e+07
Valparaíso                                   1.316022e+07
Arica y Parinacota                           1.109735e+07
Biobío                                       9.816258e+06
Atacama                                      8.323193e+06
Maule                                        7.770909e+06
Los Ríos                                     7.492851e+06
La Araucanía                                 6.944642e+06
Los Lagos                                    6.805822e+06
Magallanes y de la Antártica Chilena         6.746508e+06
Libertador General Bernardo O'Higgins        6.606043e+06
Ñuble                                        5.237524e+06
Aysén del General Carlos Ibáñez del Campo    4.201783e+06
Name: Presupuesto, dtype: float64

Ya habiendo ejecutado los codigos y obtener la presente tabla que contiene el presupuesto promedio de cada región Chilena ordenados de mayor a menor. Se puede concluir que la región que tiene el mayor presupuesto promedio es la Región Metropolitana de Santiago con $43.416.560.

## Pregunta 3

In [154]:
df_pobxpres = df_poblacion.merge(df_presupuesto)
df_pobxpres

CODIGO      MUNICIPIO  POBLACIÓN  PRESUPUESTO
0      1101        IQUIQUE     216514     59072234
1      1107  ALTO HOSPICIO     124150     13984411
2      1401   POZO ALMONTE      16683      7613962
3      1402         CAMIÑA       1345      1584008
4      1403       COLCHANE       1556      2500000
..      ...            ...        ...          ...
340   16301     SAN CARLOS      55935     10459691
341   16302       COIHUECO      28147      5001190
342   16303         ÑIQUÉN      11556      3386698
343   16304     SAN FABIÁN       4607      1798500
344   16305    SAN NICOLÁS      12068      3979700

[345 rows x 4 columns]

In [167]:
df_pobxpres = df_pobxpres[["MUNICIPIO","POBLACIÓN","PRESUPUESTO"]]
df_pobxpres

MUNICIPIO  POBLACIÓN  PRESUPUESTO
0          IQUIQUE     216514     59072234
1    ALTO HOSPICIO     124150     13984411
2     POZO ALMONTE      16683      7613962
3           CAMIÑA       1345      1584008
4         COLCHANE       1556      2500000
..             ...        ...          ...
340     SAN CARLOS      55935     10459691
341       COIHUECO      28147      5001190
342         ÑIQUÉN      11556      3386698
343     SAN FABIÁN       4607      1798500
344    SAN NICOLÁS      12068      3979700

[345 rows x 3 columns]

In [171]:
df_pobxpres.columns = ["Comuna","Poblacion","Presupuesto"]
df_pobxpres

Comuna  Poblacion  Presupuesto
0          IQUIQUE     216514     59072234
1    ALTO HOSPICIO     124150     13984411
2     POZO ALMONTE      16683      7613962
3           CAMIÑA       1345      1584008
4         COLCHANE       1556      2500000
..             ...        ...          ...
340     SAN CARLOS      55935     10459691
341       COIHUECO      28147      5001190
342         ÑIQUÉN      11556      3386698
343     SAN FABIÁN       4607      1798500
344    SAN NICOLÁS      12068      3979700

[345 rows x 3 columns]

In [180]:
df_pobxpres.sort_values(by="Poblacion", ascending=False).head(20)

Comuna  Poblacion  Presupuesto
288   PUENTE ALTO     629743     85384573
274         MAIPÚ     566664    119318265
256      SANTIAGO     486838    163383547
7     ANTOFAGASTA     410618    100635302
265    LA FLORIDA     396781     95609849
46   VIÑA DEL MAR     357228    100482380
294  SAN BERNARDO     329121     52127690
269    LAS CONDES     323309    339980450
40     VALPARAÍSO     313185     77718000
174        TEMUCO     300618     66388232
206  PUERTO MONTT     265863     52090583
277     PEÑALOLÉN     262268     56733837
78       RANCAGUA     261992     55996193
26       COQUIMBO     250947     56819394
279      PUDAHUEL     248347     87884438
25      LA SERENA     244170     59222000
280     QUILICURA     243112     46258275
320         ARICA     242510     37808221
275         ÑUÑOA     240753     63443811
141    CONCEPCIÓN     237257     44704000

In [176]:
df_pobxpres.sort_values(by= "Presupuesto", ascending=False).head(20)

Comuna  Poblacion  Presupuesto
269    LAS CONDES     323309    339980450
256      SANTIAGO     486838    163383547
278   PROVIDENCIA     154446    125041538
274         MAIPÚ     566664    119318265
7     ANTOFAGASTA     410618    100635302
46   VIÑA DEL MAR     357228    100482380
287      VITACURA      94020     97000000
265    LA FLORIDA     396781     95609849
270  LO BARNECHEA     119240     92819385
279      PUDAHUEL     248347     87884438
288   PUENTE ALTO     629743     85384573
40     VALPARAÍSO     313185     77718000
174        TEMUCO     300618     66388232
275         ÑUÑOA     240753     63443811
25      LA SERENA     244170     59222000
0         IQUIQUE     216514     59072234
26       COQUIMBO     250947     56819394
277     PEÑALOLÉN     262268     56733837
78       RANCAGUA     261992     55996193
294  SAN BERNARDO     329121     52127690

Las 20 comunas chilenas con mayor cantidad de habitantes son: Puente Alto, Maipú, Santiago, Antofagasta, La Florida, Viña del Mar, San Bernardo, Las Condes, Valparaiso, Temuco, Puerto Montt, Peñalolén, Rancagua, Coquimbo, Pudahuel, La Serena, Quilicura, Arica, Ñuñoa y Concepción.
De las cuales todas se encuentran dentro de las 20 comunas con mayor presupuesto a excepción de Puerto Montt, Quilicura, Arica y Concepción.
Con esta información y esta última observación se podría decir que a mayor población se requiere de un mayor presupuesto para sostener un bienestar social.